src/macros.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Format list-like macro invocations. These are invocations whose token trees
+// can be interpreted as expressions and separated by commas.
+// Note that these token trees do not actually have to be interpreted as
+// expressions by the compiler. An example of an invocation we would reformat is
+// foo!( x, y, z ). The token x may represent an identifier in the code, but we
+// interpreted as an expression.
+// Macro uses which are not-list like, such as bar!(key => val), will not be
+// reformated.
+// List-like invocations with parentheses will be formatted as function calls,
+// and those with brackets will be formatted as array literals.
+
+use std::thread;
+
+use syntax::ast;
+use syntax::parse::token::{Eof, Comma, Token};
+use syntax::parse::{ParseSess, tts_to_parser};
+
+use rewrite::RewriteContext;
+use expr::{rewrite_call, rewrite_array};
+use comment::FindUncommented;
+use utils::wrap_str;
+
+// FIXME: this is a pretty terrible hack. Any other solution would be preferred.
+struct ForceSend<T>(pub T);
+unsafe impl<T> Send for ForceSend<T> {}
```

Comment:
A `TokenTree` is not send, since it can contains `Rc<T>` amongst other non-`Send`s.